gapir/client/client.go

```diff
@@ -32,84 +37,195 @@ type tyLaunchArgsKey string
 // line arguments when launching GAPIR. The property must be of type []string.
 const LaunchArgsKey tyLaunchArgsKey = "gapir-launch-args"
 
+type clientInfo struct {
+	device               bind.Device
+	arch                 device.Architecture
+	abi                  *device.ABI
+	deviceConnectionInfo deviceConnectionInfo
+	connection           gapir.Connection
+	bgConnection         *backgroundConnection
+}
+
+type deviceArch struct {
+	device bind.Device
+	arch   device.Architecture
+}
+
+// ConnectionKey is used by manager to obtain a connection
+type ConnectionKey deviceArch
+
 // Client is interface used to connect to GAPIR instances on devices.
 type Client struct {
-	mutex    sync.Mutex
-	sessions map[deviceArch]*session
+	mutex       sync.Mutex
+	clientInfos map[ConnectionKey]clientInfo
 }
 
+const (
+	connectTimeout    = time.Second * 10
+	heartbeatInterval = time.Millisecond * 500
+)
+
 // New returns a newly construct Client.
 func New(ctx context.Context) *Client {
-	c := &Client{sessions: map[deviceArch]*session{}}
+	client := &Client{clientInfos: map[ConnectionKey]clientInfo{}}
 	app.AddCleanup(ctx, func() {
-		c.shutdown(ctx)
+		client.shutdown(ctx)
 	})
-	return c
-}
-
-type deviceArch struct {
-	d bind.Device
-	a device.Architecture
+	return client
 }
 
 // Connect opens a connection to the replay device.
-func (c *Client) Connect(ctx context.Context, d bind.Device, abi *device.ABI) (gapir.Connection, error) {
+func (client *Client) Connect(ctx context.Context, device bind.Device, abi *device.ABI) (*ConnectionKey, error) {
+	client.mutex.Lock()
+	defer client.mutex.Unlock()
```

Comment:
Concurrency is never easy: can you add a small comment at the top of the file, saying why these internal functions needs to be thread-safe?